### PR TITLE
DAOS-13144 test: Enable valgrind on tests that use NVMe

### PR DESCRIPTION
**Pull request by jolivier23**
Split NVMe tests into new stages, one for valgrind and one for
memcheck since we don't require VMs with bdev for all unit tests.

Move unitTestPost function from pipeline-lib to Jenkinsfile
and rename to unitTestPostEx so the pipeline-lib doesn't have
to change immediately.  I needed to make changes in the
function and it shouldn't be in pipeline-lib anyway, though
possibly in a local library, if that's possible.

Enable valgrind on tests that require sudo.  Change the
--no-sudo unit test argument so it's more flexible.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -20,6 +20,117 @@
 /* groovylint-disable-next-line CompileStatic */
 job_status_internal = [:]
 
+/**
```
Comment (jolivier23) [inline Jenkinsfile:23]: I needed to edit the unitTestPost function so I just moved the whole thing as it belongs in the DAOS repo anyway and doesn't affect the bytecode limit.

File: Jenkinsfile [hunk 2]
```diff
@@ -729,6 +840,28 @@ pipeline {
...
+                        job_step_update(
+                            unitTest(timeout_time: 60,
+                                     unstash_opt: true,
+                                     inst_repos: prRepos(),
+                                     inst_rpms: unitPackages()))
+                    }
+                    post {
+                        always {
+                            unitTestPostEx(['unit_test_logs/'])
+                            job_status_update()
```
Comment (JohnMalmberg) [inline Jenkinsfile:860]: A pipeline-lib routine can not call job_status_update due to variable scope.
The unitTestPostEx should be able to call it, which would simplify the stages.

File: Jenkinsfile [hunk 3]
```diff
@@ -20,6 +20,117 @@
...
+    if (ignore_failure) {
+      health_scale = 0.0
+    }
+
+    def cb_result = currentBuild.result
+    junit testResults: testResults,
+          healthScaleFactor: health_scale
+
+    if (cb_result != currentBuild.result) {
+      println "The junit plugin changed result to ${currentBuild.result}."
```
Comment (JohnMalmberg) [inline Jenkinsfile:72]: We have a hidden wrapper for junit that does the cb_result check and logs the result, so the cb_result check here is not needed.

File: Jenkinsfile [hunk 4]
```diff
@@ -20,6 +20,117 @@
...
+                                      targetLocation: target_dir)])
+    sh "tar -czf ${tgz_file} ${target_dir}"
+  }
+
+  artifacts.each {
+    archiveArtifacts artifacts: it,
+                     allowEmptyArchive: ignore_failure
+  }
+
+  def target_stash = "${stage_info['target']}-${stage_info['compiler']}"
```
Comment (JohnMalmberg) [inline Jenkinsfile:92]: String target_stash

File: Jenkinsfile [hunk 5]
```diff
@@ -20,6 +20,117 @@
...
+  if (stage_info['compiler'] == 'covc') {
+    return
+  }
+
+  if (valgrind_stash) {
+    stash name: valgrind_stash, includes: valgrind_pattern
+  }
+
+  if (stage_info['NLT']) {
+    def cb_result = currentBuild.result
```
Comment (JohnMalmberg) [inline Jenkinsfile:107]: String cb_result

File: Jenkinsfile (hunk 3 above)
Comment (jolivier23) [reply Jenkinsfile:72]: This is copied from pipeline-lib and I'd like to limit changes until subsequent patches.

File: Jenkinsfile (hunk 5 above)
Comment (jolivier23) [reply Jenkinsfile:108]: See above

File: Jenkinsfile (hunk 4 above)
Comment (jolivier23) [reply Jenkinsfile]: this one should be fixed.

File: Jenkinsfile (hunk 2 above)
Comment (jolivier23) [reply Jenkinsfile]: I made this change to test.  We will need to add a ccov stage at some point for the bdev tests so every little bit can help.

**Review by JohnMalmberg — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Pull request merged**